Sources/SpeziViews/Views/ManagedNavigationStack/NavigationStepIdentifier.swift

```diff
@@ -0,0 +1,114 @@
+            case (.viewTypeAndSourceLoc, .identifiable), (.identifiable, .viewTypeAndSourceLoc):
+                false
+            }
+        }
+    }
+    
+    let identifierKind: IdentifierKind
+    let viewType: any View.Type
+    let flowElementSourceLocation: _NavigationFlow.Element.SourceLocation?
+    
+    /// Whether the step is custom, i.e. not one of the steps defined via the ``NavigationFlowBuilder`` but instead created via e.g. ``ManagedNavigationStack/Path/append(customView:)``.
+    var isCustom: Bool {
+        flowElementSourceLocation == nil
+    }
+    
+    /// Initializes an identifier using a view. If the view conforms to `Identifiable`, its `id` is used; otherwise, the view's type is used.
+    /// - Parameters:
+    ///   - view: The view used to initialize the identifier.
+    ///   - custom: A flag indicating whether the step is custom.
+    @MainActor
+    init(element: _NavigationFlow.Element) {
+        self.viewType = type(of: element.view)
+        self.flowElementSourceLocation = element.sourceLocation
+        if let identifiable = element.view as? any NavigationStepIdentifiable {
+            let id = identifiable.id
+            self.identifierKind = .identifiable(id)
+        } else if let identifiable = element.view as? any Identifiable {
+            let id = identifiable.id
+            self.identifierKind = .identifiable(id)
+        } else {
+            self.identifierKind = .viewTypeAndSourceLoc
+        }
+    }
+}
+
+
+extension NavigationStepIdentifier: Hashable {
+    static func == (lhs: Self, rhs: Self) -> Bool {
+        switch (lhs.identifierKind, rhs.identifierKind) {
+        case (.viewTypeAndSourceLoc, .viewTypeAndSourceLoc):
+            lhs.viewType == rhs.viewType && lhs.flowElementSourceLocation == rhs.flowElementSourceLocation
+        case let (.identifiable(lhsValue), .identifiable(rhsValue)):
+            lhsValue.isEqual(rhsValue)
+        case (.viewTypeAndSourceLoc, .identifiable), (.identifiable, .viewTypeAndSourceLoc):
+            false
+        }
+    }
+    
+    func hash(into hasher: inout Hasher) {
+        switch self.identifierKind {
+        case .viewTypeAndSourceLoc:
+            hasher.combine(ObjectIdentifier(viewType))
+            if let flowElementSourceLocation {
+                hasher.combine(flowElementSourceLocation)
+            }
+        case .identifiable(let value):
+            hasher.combine(ObjectIdentifier(type(of: value)))
+            hasher.combine(value)
+        }
+    }
+}
+
+
+extension NavigationStepIdentifier: CustomDebugStringConvertible {
+    var debugDescription: String {
+        var desc = "\(Self.self)(isCustom: \(isCustom), viewType: \(viewType), identifierKind: \(identifierKind)"
+        if let sourceLoc = flowElementSourceLocation {
+            desc += ", sourceLoc: \(sourceLoc.fileId);\(sourceLoc.line);\(sourceLoc.column)"
+        }
+        desc += ")"
+        return desc
+    }
+}
+
+
+extension Equatable {
```

Comment:
Do we really need that? Super ugly imo, maybe we can refactor this somehow (even leading to some code duplication is fine imo)

Reply:
we do, since that's how we compare the identities of steps with custom identifiers. the previous version was even uglier; it manually computed the hash value of the identifier and then stored it as an untyped `Int`. this approach at least operates on the identifier values directly (which is also more accurate, since we're no longer treating 2 different identifiers as equal if they had identical hash values, and we're also taking the type of the identifier into account)

Reply:
Thanks for clarifying!